stdlib/asm/collections/smt.masm

```diff
@@ -1,1436 +1,315 @@
-    # first 4 flags onto the operand stack; meaning of the flags f0, f1, and f2 depends
-    # on what type of insert is being executed (4 cycles)
-    adv.push_smtset adv_push.4
-    # => [is_update, f0, f1, f2, V, K, R, ...]
-
-    # execute the actual insert procedure
-    exec.insert_internal
-    # => [V_old, R_new, ...]
-end
-
-# DELETE
+# GET
 # =================================================================================================
 
-#! Verifies that a node at depth 16 and index defined by the most significant element of K in a
-#! tree with root R is a root of an empty subtree.
-#!
-#! Input:  [Z, K, R, ...]
-#! Output: [Z, R, ...]
+#! Returns the value located under the specified key in the Sparse Merkle Tree defined by the
+#! specified root.
 #!
-#! Where Z is [ZERO; 4].
-#!
-#! Cycles: 26
-proc.verify_empty_node_16
-    # (9 cycles)
-    swapw.2 dup.4 exec.get_top_16_bits push.16
-    # => [16, idx, R, K, Z, ...]
-
-    # (4 cycles)
-    push.EMPTY_16_0.EMPTY_16_1.EMPTY_16_2.EMPTY_16_3
-    # => [E16, 16, idx, R, K, Z, ...]
-
-    # (1 cycle)
-    mtree_verify
-    # => [E16, 16, idx, R, K, Z, ...]
-
-    # (12 cycles)
-    dropw drop drop swapw dropw swapw
-    # => [Z, R, ...]
-end
-
-#! Verifies that a node at depth 32 and index defined by the most significant element of K in a
-#! tree with root R is a root of an empty subtree.
+#! If no values had been previously inserted under the specified key, an empty word (i.e.,
+#! [ZERO; 4]) is returned.
 #!
-#! Input:  [Z, K, R, ...]
-#! Output: [Z, R, ...]
+#! Inputs:
+#!   Operand stack: [K, R, ...]
 #!
-#! Where Z is [ZERO; 4].
+#! Outputs:
+#!   Operand stack: [V, R, ...]
 #!
-#! Cycles: 23
-proc.verify_empty_node_32
-    # (6 cycles)
-    swapw.2 dup.4 exec.get_top_32_bits push.32
-    # => [32, idx, R, K, Z, ...]
-
-    # (4 cycles)
-    push.EMPTY_32_0.EMPTY_32_1.EMPTY_32_2.EMPTY_32_3
-    # => [E32, 32, idx, R, K, Z, ...]
-
-    # (1 cycle)
-    mtree_verify
-    # => [E32, 32, idx, R, K, Z, ...]
-
-    # (12 cycles)
-    dropw drop drop swapw dropw swapw
-    # => [Z, K, ...]
-end
-
-#! Verifies that a node at depth 48 and index defined by the most significant element of K in a
-#! tree with root R is a root of an empty subtree.
-#!
-#! Input:  [Z, K, R, ...]
-#! Output: [Z, R, ...]
-#!
-#! Where Z is [ZERO; 4].
-#!
-#! Cycles: 29
-proc.verify_empty_node_48
-    # (12 cycles)
-    swapw.2 dup.4 exec.get_top_48_bits push.48
-    # => [48, idx, R, K, Z, ...]
-
-    # (4 cycles)
-    push.EMPTY_48_0.EMPTY_48_1.EMPTY_48_2.EMPTY_48_3
-    # => [E48, 48, idx, R, K, Z, ...]
-
-    # (1 cycle)
-    mtree_verify
-    # => [E48, 48, idx, R, K, Z, ...]
-
-    # (12 cycles)
-    dropw drop drop swapw dropw swapw
-    # => [Z, K, ...]
-end
+#! Fails if the tree with the specified root does not exits in the VM's advice provider.
+export.get
+    # Prepare for `mtree_get`
+    dupw.1 dup.4 push.LEAF_DEPTH
+    # => [depth, K[0], R, K, R]
```

Comment:
I struggled to properly described this. `K[3]` is true when the key is accessed from Rust code. But since we revert the order of words when pushing on the stack, the most significant element becomes the first (i.e. `WORD[0]`). 

That being said, I went for `K[0]` since it maps better with what the MASM is doing. That is, we do `dup.4` (to duplicate the first element of second word on stack) as opposed to `dup.7` (which would duplicate the last element - `WORD[3]` so to speak).

However this change of "coordinate frame" from MASM to Rust is confusing, and I'm not sure if this is the right approach?

Reply:
The convention we've been using is that a word is laid out on the stack in reverse order (i.e., most significant element closer to the top of the stack). So, the following two are equivalent:
```
[WORD, ...]
[WORD[3], WORD[2], WORD[1], WORD[0], ...]
```
With this convention in mind, I think all aligns correctly - the way the words are represented in Rust, the way the are stored in VM memory, and the way they appear on the stack.

Reply:
Changed to `K[3]`